### ISSUE TITLE
torch.clamp's boundary gradient changed between 2.9.1 and 2.14.0, so kornia's clamp-as-gradient-guard pattern silently fails on the older half of its supported range

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4229]
## Summary

`torch.clamp`'s gradient **at the clamp bound itself** changed between torch 2.9.1 and 2.14.0: it used to pass
the incoming gradient through (`1.0`), and now returns `0.0`. kornia declares `torch>=2.5.1` and its PR CI runs
**three** torch versions (2.5.1 and 2.14.0 in the main matrix, 2.9.1 on the half-precision legs), so the two ends
of the supported range disagree.

That matters because kornia repeatedly uses `clamp` as a *gradient* guard directly in front of an operator whose
own derivative is unbounded exactly at the clamp bound — `clamp(min=0.0).sqrt()`, `clamp(-1, 1).acos()`,
`clamp(-1, 1).asin()`. Wherever the bound **is** the singular point, the guard silently works on 2.14 and does
**not** work on 2.5.1 / 2.9.1: the clamp hands the unbounded derivative straight through and the result is `inf`
or `nan`.

Measured, same machine, CPU:

| | torch 2.5.1 | torch 2.9.1 | torch 2.14.0 |
| --- | --- | --- | --- |
| `clamp(x, min=0.0)` grad at `x = 0` | `1.0` | `1.0` | **`0.0`** |
| `clamp(w, -1, 1)` grad at `w = 1` | `1.0` | `1.0` | **`0.0`** |
| `clamp(x, min=0.0).sqrt()` grad at `x = 0` | `inf` | `inf` | `0.0` |
| `clamp(w, -1, 1).acos()` grad at `w = 1` | `-inf` | `-inf` | `0.0` |
| bare `acos` grad at `w = 1` | `-inf` | `-inf` | `-inf` |
| bare `asin` grad at `x = 1` | `inf` | `inf` | `inf` |

The last two rows are there to head off the obvious wrong diagnosis: **`acos`/`asin`/`sqrt` did not change**.
Only `clamp`'s treatment of the closed boundary did.

## A reachable instance: NaN gradients from half-precision descriptor matching

`kornia/feature/matching.py:51`

```python
dm = d1_sq.repeat(1, d2.size(0)) + d2_sq.repeat(1, d1.size(0)).t() - 2.0 * d1 @ d2.t()
dm = dm.clamp(min=0.0).sqrt()
```

This is `_cdist`'s fallback branch, taken for **MPS tensors, `float16`/`bfloat16`, and under export** (the
`torch.cdist` fast path covers the rest). An exactly-zero distance — two identical descriptors — puts `dm` on the
clamp bound.

```python
import torch
from kornia.feature.matching import _cdist

for dt in (torch.float16, torch.float32):
    d1 = torch.tensor([[1.0, 0.0], [0.0, 1.0]], dtype=dt, requires_grad=True)
    d2 = torch.tensor([[1.0, 0.0], [2.0, 2.0]], dtype=dt)   # row 0 identical to d1 row 0
    dm = _cdist(d1, d2)
    dm.sum().backward()
    print(dt, dm[0, 0].item(), d1.grad.flatten().tolist())
```

```
torch 2.5.1    float16   0.0   [nan, nan, -1.6015625, 0.259765625]
torch 2.5.1    float32   0.0   [-0.44721359, -0.89442718, -1.60153389, 0.25989318]
torch 2.14.0   float16   0.0   [-0.447265625, -0.89453125, -1.6015625, 0.259765625]
torch 2.14.0   float32   0.0   [-0.44721359, -0.89442718, -1.60153389, 0.25989318]
```

`float32` on CPU takes `torch.cdist` and is finite on both. `float16` takes the clamp branch and is **`nan` on
2.5.1** and finite on 2.14. Note this is the branch with the least test coverage in the repo, and the
half-precision CI legs pin **2.9.1** — the version where it is broken.

Zero distances are not exotic here: kornia's detectors pad their output with zero LAFs, and identical padded
descriptors match at ~0 distance (this is the mechanism behind the padding self-match problem already known in
`kornia.feature`).

## Audit of the pattern

`grep -rnE "clamp[^)]*\)\.(acos|asin|atanh|sqrt|log|rsqrt)\(|(acos|asin|atanh|sqrt|log|rsqrt)\(.*clamp" kornia/`

**Bound is the singular point — affected:**

- `kornia/feature/matching.py:51` — `dm.clamp(min=0.0).sqrt()` (reproduced above)
- `kornia/geometry/solvers/polynomial_solver.py:284` — `torch.sqrt(torch.clamp(R_sq, min=0.0))`
- `kornia/geometry/solvers/polynomial_solver.py:299` — `torch.sqrt(torch.clamp(radicand, min=0.0))`
- `kornia/geometry/conversions.py:1156` — `acos(clamp(w, -1, 1))`, which is #4007 and which #4228 is fixing

**Bound is a positive floor — safe on every version, no action:**

- `kornia/feature/aliked/aliked.py:171` (`min=1e-8`), `kornia/geometry/conversions.py:473` (`min=1e-12`),
  `kornia/geometry/liegroup/se3.py:201` (`clamp_min(1e-12)`),
  `kornia/geometry/camera/distortion_kannala_brandt.py:148` (`min=1e-16`),
  `kornia/feature/loftr/utils/fine_matching.py:83` (`min=1e-10`).

I only reproduced an end-to-end NaN for `matching.py`; the two `polynomial_solver.py` sites are listed because
they match the pattern, not because I found an input that reaches them (`solve_cubic` on a degenerate cubic came
back finite on both versions).

## Suggested scope

1. **Do not fix these by clamping.** A clamp whose bound is the singularity is not a gradient guard on the
   majority of kornia's supported torch versions. The working pattern is the one #4228 uses: substitute a safe
   argument into the call that gets *differentiated*, and take the value from a `.detach()`ed copy of the real
   argument — that is version-independent by construction.
2. Fix `matching.py:51`, with a pin that differentiates through `_cdist` at an exact zero distance in
   `float16`/`bfloat16` (a `float32` CPU pin proves nothing — it takes the other branch).
3. Decide whether the two `polynomial_solver.py` sites are reachable at their bound and either fix or leave them
   with a note.
4. Worth a line in `AGENTS.md`'s precision section: `clamp` is not a portable gradient guard across kornia's
   supported torch range.

## Environment

kornia `main` @ `b56a1bc2`, torch 2.5.1 / 2.9.1 / 2.14.0, python 3.11, macOS arm64, CPU.
Related: #4007 (one instance), #4228 (fixes that instance), #4203 (the same "two ends of the supported torch
range disagree" shape, for `inv_ex`).

Posted on behalf of @ducha-aiki by Claude (Claude Opus 5).

**Comment by cipher-v** [comment]
Hi @ducha-aiki, I’d like to work on this. My initial scope would be the _cdist fallback in kornia/feature/matching.py: implement the version-independent safe-argument/detached-value pattern and add a focused CPU float16 gradient regression test. Should the polynomial-solver audit be included in the same PR, or handled separately? If this is available, please assign it to me.
